pxtlib/auth.ts

```diff
@@ -312,61 +312,96 @@ namespace pxt.auth {
-            if (!ops.length) { return {success: true, res: curPref} }
-            ts.pxtc.jsonPatch.patchInPlace(curPref, ops);
-            await this.setUserPreferencesAsync(curPref);
-            // If we're not logged in, non-persistent local state is all we'll use
-            if (!await this.loggedInAsync()) { return {success: true, res: curPref}; }
-
-            // If the user is logged in, save to cloud, but debounce the api call as this can be called frequently from skillmaps
-
-            // Replace matching patches in the queue
-            ops.forEach((incoming, iIncoming) => {
-                this.prefPatchOps.some((existing, iExisting) => {
-                    if (!ts.pxtc.jsonPatch.opsAreEqual(existing, incoming)) return false;
-                    // Patches are equivalent, replace in queue
-                    this.prefPatchOps[iExisting] = incoming;
-                    // Clear from incoming so we don't add it below
-                    (ops as ts.pxtc.jsonPatch.PatchOperation[])[iIncoming] = null;
-                    return true;
-                });
-            });
-            // Add remaining ops to the queue
-            ops.filter(op => !!op).forEach(op => this.prefPatchOps.push(op));
+        private patchQueue: {
+            ops: ts.pxtc.jsonPatch.PatchOperation[];
+            filter?: (op: ts.pxtc.jsonPatch.PatchOperation) => boolean;
+        }[] = [];
+
+        public async patchUserPreferencesAsync(patchOps: ts.pxtc.jsonPatch.PatchOperation | ts.pxtc.jsonPatch.PatchOperation[], opts: {
+            immediate?: boolean,                                            // Sync the change with cloud immediately, don't debounce
+            filter?: (op: ts.pxtc.jsonPatch.PatchOperation) => boolean      // Filter the final set of patch operations
+        } = {}): Promise<SetPrefResult> {
+            const defaultSuccessAsync = async (): Promise<SetPrefResult> => ({ success: true, res: await this.userPreferencesAsync() });
+
+            patchOps = Array.isArray(patchOps) ? patchOps : [patchOps];
+            patchOps = patchOps.filter(op => !!op);
+            if (!patchOps.length) { return await defaultSuccessAsync(); }
+
+            const patchDiff = (pSrc: UserPreferences, ops: ts.pxtc.jsonPatch.PatchOperation[], filter?: (op: ts.pxtc.jsonPatch.PatchOperation) => boolean) => {
+                // Apply patches to pDst and return the diff as a set of new patch ops.
+                const pDst = U.deepCopy(pSrc);
+                ts.pxtc.jsonPatch.patchInPlace(pDst, ops);
+                let diff = ts.pxtc.jsonPatch.diff(pSrc, pDst);
+                // Run caller-provided filter
+                if (diff.length && filter) { diff = diff.filter(filter); }
+                return diff;
+            }
+
+            // Process incoming patch operations to produce a more fine-grained set of diffs. Incoming patches may be overly destructive
+            {
```

Comment:
I might be super confused, but I can't figure out what these braces are for????

Reply:
Just scoping the variables. I use `diff` and `curPrefs` a lot in this function.